c/tskit/trees.c

```diff
@@ -2012,6 +2018,113 @@ tsk_treeseq_trait_correlation(tsk_treeseq_t *self,
     return ret;
 }
 
+static int
+trait_regression_summary_func(size_t state_dim, double *state, size_t result_dim,
+        double *result, void *params)
+{
+    covariates_stat_params_t args = *(covariates_stat_params_t *) params;
+    const double num_samples = (double) args.num_samples;
+    const tsk_size_t k = args.num_covariates;
+    const double *V = args.V;;
+    const double *x = state;
+    const double *v;
+    double m, a, denom, xx;
+    size_t i, j;
+
+    m = x[state_dim - 1];
+    for (i = 0; i < result_dim; i++) {
+        if ((m > 0.0) && (m < num_samples)) {
+            v = GET_2D_ROW(V, k, i);
+            a = x[i];
+            denom = m;
+            for (j = 0; j < k; j++) {
+                xx = x[result_dim + j];
+                a -= xx * v[j];
+                denom -= xx * xx;
+            }
+            if (denom < 1e-8) {
```

Comment:
Hmmm; feels very arbitrary. We need a comment describing why this is done at the least. Should it be a tuning parameter, passed in to the top-level method?

Reply:
Maybe this should be a `tol` parameter. In practice I was getting values of like 1e-15 here, so something like that would probably be fine, so this is rather conservative. 

But: denom is equal to the norm of the orthogonal projection of the genotype vector into the subspace spanned by the covariates.  A value of `denom` less than 1e-8 here means that the genotype is *very* close to being exactly in the span of the covariates.  For denom to be of size 1e-8 but the genotype *not* be in the span of the covariates, given that the genotype vector is binary, you'd need something like, say, sex as a covariate and a SNP that exactly segregates between sexes except for in one person out of a sample size of 100 million. So, yeah. `denom` is only anywhere close to zero in in either test data, where we've got like four samples and four covariates, or in cases where the genotype is very tightly correlated with a covariate, someplace that it's probably good to be conservative.

So: I'll add a comment?

Reply:
What happens if we just let it pass through though? Do things blow up, or do we just get something quite large (I guess, since we're dividing by 1e-16?). What's the harm in getting back these values?

If we have to have a threshold for numerical reasons then we'll have to add it as a parameter in the C API; unlike Python, we can't just add new parameters later, unfortunately. If we hard code it in, in practise we'd end up having to add a second version of this function in with this parameter included.

Reply:
Actually, in practise we seem to get not-large values, because the numerator is small also. We don't want these because (a) we can't test them and (b) if the denominator is zero then the solution is non-unique, and the one we want to return here is 0.

I suppose we could add this as a parameter to the C API, although I really don't think anyone's going to want to change it, ever. I'm still voting to leave it out.

Reply:
OK, let's leave it as it is then.